### PR TITLE
Fix Ruby/Python SWIG API when type is AlfalfaComponent

### PR DESCRIPTION
**Pull request by TShapinsky**
Pull request overview
---------------------

Add additional function in SWIG for `AlfalfaJSON::exposeFromComponent`, `AlfalfaPoint::setInput` and `AlfalfaPoint::setOutput` which provides a method which takes `const AlfalfaComponentBase&` instead of `AlfalfaComponent`. This is required because it seems like SWIG cannot understand the converting constructor.

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/alfalfa/Alfalfa.i [hunk 1]
```diff
@@ -57,4 +57,26 @@
...
+namespace openstudio::alfalfa {
+  %extend AlfalfaPoint {
+    void setOutput(const AlfalfaComponentBase& component) {
+      AlfalfaComponent alfalfa_component(component);
+      $self->setOutput(alfalfa_component);
+    }
+
+    void setInput(const AlfalfaComponentBase& component) {
+      AlfalfaComponent alfalfa_component(component);
+      $self->setInput(alfalfa_component);
```
Comment (TShapinsky) [inline src/alfalfa/Alfalfa.i:69]:
```suggestion
      $self->setInput(AlfalfaComponent {component});
```

File: src/alfalfa/Alfalfa.i [hunk 2]
```diff
@@ -57,4 +57,26 @@
 %template(OptionalAlfalfaPoint) boost::optional<openstudio::alfalfa::AlfalfaPoint>;
 %template(OptionalAlfalfaComponent) boost::optional<openstudio::alfalfa::AlfalfaComponent>;
 
+namespace openstudio::alfalfa {
+  %extend AlfalfaPoint {
+    void setOutput(const AlfalfaComponentBase& component) {
+      AlfalfaComponent alfalfa_component(component);
+      $self->setOutput(alfalfa_component);
```
Comment (TShapinsky) [inline src/alfalfa/Alfalfa.i:64]:
```suggestion
      $self->setInput(AlfalfaComponent {component});
```

File: src/alfalfa/Alfalfa.i [hunk 3]
```diff
@@ -57,4 +57,26 @@
...
+    void setInput(const AlfalfaComponentBase& component) {
+      AlfalfaComponent alfalfa_component(component);
+      $self->setInput(alfalfa_component);
+    }
+  }
+
+  %extend AlfalfaJSON {
+    boost::optional<AlfalfaPoint> AlfalfaJSON::exposeFromComponent(const AlfalfaComponentBase& component, const std::string& display_name = std::string()) {
+      AlfalfaComponent alfalfa_component(component);
+      return $self->exposeFromComponent(alfalfa_component, display_name);
```
Comment (TShapinsky) [inline src/alfalfa/Alfalfa.i:76]:
```suggestion
      return $self->exposeFromComponent(AlfalfaComponent {component}, display_name);
```

File: src/alfalfa/Alfalfa.i [hunk 4]
```diff
@@ -57,4 +57,26 @@
 %template(OptionalAlfalfaPoint) boost::optional<openstudio::alfalfa::AlfalfaPoint>;
 %template(OptionalAlfalfaComponent) boost::optional<openstudio::alfalfa::AlfalfaComponent>;
 
+namespace openstudio::alfalfa {
+  %extend AlfalfaPoint {
```
Comment (TShapinsky) [inline src/alfalfa/Alfalfa.i:61]: These could be moved into the main source if desired.

File: src/alfalfa/AlfalfaComponent.hpp [hunk 5]
```diff
@@ -12,8 +12,7 @@ namespace alfalfa {
   class ALFALFA_API AlfalfaComponent
   {
    public:
-    template <typename T, std::enable_if_t<std::is_base_of<AlfalfaComponentBase, T>::value, bool> = true>
-    AlfalfaComponent(T component) : m_component(std::make_unique<T>(std::move(component))) {}
+    AlfalfaComponent(const AlfalfaComponentBase& component) : m_component(component.clone()) {}
```
Comment (jmarrec) [inline src/alfalfa/AlfalfaComponent.hpp:15]: Curious why you added this one in the main source but the others just in the Alfalfa.i

**Review by jmarrec — APPROVED** [review]
I'm liking this AlfalfaComponent / AlfalfaComponentBase template stuff less and less. But this appears to work fine.

File: src/alfalfa/AlfalfaComponent.hpp (hunk 5 above)
Comment (TShapinsky) [reply src/alfalfa/AlfalfaComponent.hpp:15]: I didn't want the ctor exposed to the bindings so i added it here and ignored the ctors for the class in the .i .

**Comment by TShapinsky** [comment]
> I'm liking this AlfalfaComponent / AlfalfaComponentBase template stuff less and less. But this appears to work fine.

Agreed. If i get some more time before the next release it's something I want to take a look at again.

**Comment by jmarrec** [comment]
@wenyikuang where do we merge this one?

**Comment by wenyikuang** [comment]
> @wenyikuang where do we merge this one?

I haven't merged it yet, maybe in the final?

**Comment by jmarrec** [comment]
@wenyikuang I'm asking because it seems that master is well ahead of develop

**Pull request merged**